numpy/core/src/multiarray/convert.c

```diff
@@ -302,57 +302,34 @@ PyArray_ToString(PyArrayObject *self, NPY_ORDER order)
     npy_intp i;
     char *dptr;
     int elsize;
-    PyObject *ret;
+    PyObject *ravel_out, *ret;
     PyArrayIterObject *it;
 
-    if (order == NPY_ANYORDER)
-        order = PyArray_ISFORTRAN(self);
-
-    /*        if (PyArray_TYPE(self) == NPY_OBJECT) {
-              PyErr_SetString(PyExc_ValueError, "a string for the data" \
-              "in an object array is not appropriate");
-              return NULL;
-              }
-    */
-
     numbytes = PyArray_NBYTES(self);
-    if ((PyArray_IS_C_CONTIGUOUS(self) && (order == NPY_CORDER))
-        || (PyArray_IS_F_CONTIGUOUS(self) && (order == NPY_FORTRANORDER))) {
-        ret = PyBytes_FromStringAndSize(PyArray_DATA(self), (Py_ssize_t) numbytes);
+
+    ravel_out = PyArray_Ravel(self, order);
+    it = (PyArrayIterObject *)PyArray_IterNew(ravel_out);
+    if (ravel_out == NULL) {
+        Py_DECREF(it);
+        return NULL;
     }
-    else {
-        PyObject *new;
-        if (order == NPY_FORTRANORDER) {
-            /* iterators are always in C-order */
-            new = PyArray_Transpose(self, NULL);
-            if (new == NULL) {
-                return NULL;
-            }
-        }
-        else {
-            Py_INCREF(self);
-            new = (PyObject *)self;
-        }
-        it = (PyArrayIterObject *)PyArray_IterNew(new);
-        Py_DECREF(new);
-        if (it == NULL) {
-            return NULL;
-        }
-        ret = PyBytes_FromStringAndSize(NULL, (Py_ssize_t) numbytes);
-        if (ret == NULL) {
-            Py_DECREF(it);
-            return NULL;
-        }
-        dptr = PyBytes_AS_STRING(ret);
-        i = it->size;
-        elsize = PyArray_DESCR(self)->elsize;
-        while (i--) {
-            memcpy(dptr, it->dataptr, elsize);
-            dptr += elsize;
-            PyArray_ITER_NEXT(it);
-        }
+    Py_DECREF(ravel_out);
+
+    ret = PyBytes_FromStringAndSize(NULL, (Py_ssize_t) numbytes);
+    if (ret == NULL) {
```

Comment:
not sure if we need this here just as a micro opt or if there is a deeper reason ?